Components/Sensors/MTK3339.cs

```diff
@@ -29,95 +23,75 @@ public class MTK3339 : ISensor
-        {
-            if (!GettingHasFix)
+            while (true)
             {
-                Thread T = new Thread(() =>
+                char Data = BlockingGetChar();
+                Builder.Append(Data);
+                if (Data == '\n')
                 {
-                    GettingHasFix = true;
-                    OnFinish.Invoke(HasFix());
-                    GettingHasFix = false;
-                });
-                T.Start();
+                    if (Builder.ToString().Contains("$GPGGA"))
+                    {
+                        int? LatIndex = null;
+                        int? LngIndex = null;
+                        string[] SplitData = Builder.ToString().Split(',');
+                        for (int i = 0; i < SplitData.Length; i++)
+                        {
+                            switch (SplitData[i])
+                            {
+                                case "N": LatIndex = i - 1; break;
+                                case "S": LatIndex = -(i - 1); break;
+                                case "W": LngIndex = -(i - 1); break;
+                                case "E": LngIndex = i - 1; break;
+                            }
+                        }
+
+                        if (LatIndex.HasValue && LngIndex.HasValue)
+                        {
+                            Latitude = RawToDeg(SplitData[Math.Abs(LatIndex.Value)]) * Math.Sign(LatIndex.Value);
+                            Longitude = RawToDeg(SplitData[Math.Abs(LngIndex.Value)]) * Math.Sign(LngIndex.Value);
+                        }
+                    }
+                    Builder.Clear();
+                }
             }
         }
 
-        public void GetCoords(Action<Tuple<float, float>> OnFinish)
+        private char BlockingGetChar()
         {
-            if (!GettingCoords)
-            {
-                Thread T = new Thread(() =>
-                {
-                    GettingCoords = true;
-                    OnFinish.Invoke(GetCoords());
-                    GettingCoords = false;
-                });
-                T.Start();
-            }
+            byte[] NewChar = new byte[1];
+            while (UART.BytesAvailable() < 1) ;
+            UART.Read(1, NewChar);
+            return (char)NewChar[0];
         }
 
-        /// <summary> Reads an NMEA sentence and splits it. </summary>
-        /// <returns> Returns the NMEA sentence split by comma. </returns>
-        private string[] Read()
+        /// <summary> Checks whether this GPS has a fix. </summary>        
+        /// <returns> Returns true if the GPS has a fix and false otherwise. </returns>
+        public bool Test() => HasFix();
+
+        /// <summary> Gets new readings from GPS. </summary>
+        public void UpdateState() => GetCoords();
+
+        /// <summary> Queries the GPS to see if it has a fix. </summary>
+        /// <returns> <c>true</c>, if fix was hased, <c>false</c> otherwise. </returns>
+        public bool HasFix()
```

Comment:
Another issue I saw when editing. This is not correct at all. The GPS definitely does not always have a fix. Either remove this method, or implement the actual functionality you're promising.